pkg/store/gcs.go

```diff
@@ -581,60 +581,154 @@ func (b *gcsBlock) Close() error {
 }
 
 type gcsChunkReader struct {
+	logger log.Logger
 	ctx    context.Context
-	cancel func()
+	close  func()
 	id     ulid.ULID
 
-	files []*storage.ObjectHandle
+	files    []*storage.ObjectHandle
+	preloads [][]uint32
+	mtx      sync.Mutex
+	chunks   map[uint64]chunks.Chunk
 }
 
-func newGCSChunkReader(ctx context.Context, id ulid.ULID, files []*storage.ObjectHandle) *gcsChunkReader {
+func newGCSChunkReader(ctx context.Context, logger log.Logger, id ulid.ULID, files []*storage.ObjectHandle, close func()) *gcsChunkReader {
 	ctx, cancel := context.WithCancel(ctx)
+
 	return &gcsChunkReader{
-		ctx:    ctx,
-		cancel: cancel,
-		id:     id,
-		files:  files,
+		logger:   logger,
+		ctx:      ctx,
+		id:       id,
+		files:    files,
+		preloads: make([][]uint32, len(files)),
+		chunks:   map[uint64]chunks.Chunk{},
+		close: func() {
+			cancel()
+			close()
+		},
 	}
 }
 
-func (r *gcsChunkReader) Chunk(id uint64) (chunks.Chunk, error) {
+// addPreload adds the chunk with id to the data set that will be fetched on calling preload.
+func (r *gcsChunkReader) addPreload(id uint64) error {
 	var (
 		seq = int(id >> 32)
-		off = int((id << 32) >> 32)
+		off = uint32(id)
 	)
-	if seq >= len(r.files) {
-		return nil, errors.Errorf("reference sequence %d out of range", seq)
+	if seq >= len(r.preloads) {
+		return errors.Errorf("reference sequence %d out of range", seq)
 	}
-	obj := r.files[seq]
+	r.preloads[seq] = append(r.preloads[seq], off)
+	return nil
+}
 
-	// We don't know the length of the chunk, load 1024 bytes, which should be enough
-	// even for long ones.
-	rd, err := obj.NewRangeReader(r.ctx, int64(off), int64(off+1024))
-	if err != nil {
-		return nil, errors.Wrap(err, "create reader")
-	}
-	defer rd.Close()
+// preloadFile adds actors to load all chunks referenced by the offsets from the given file.
+// It attempts to conslidate requests for multiple chunks into a single one and populates
+// the reader's chunk map.
+func (r *gcsChunkReader) preloadFile(g *run.Group, seq int, file *storage.ObjectHandle, offsets []uint32) {
+	sort.Slice(offsets, func(i, j int) bool {
+		return offsets[i] < offsets[j]
+	})
+	j := 0
+	k := 0
+
+	for k < len(offsets) {
+		j = k
+		k++
+
+		start := offsets[j]
+		end := start + 1024
```

Comment:
Can you name all these constants ? `1024`

Reply:
Why exactly we are crawling every 1KB here? so... maybe just comment here?

Reply:
From the index alone we don't know how long the chunk will actually be. But we can basically tell that they'll never really be more than 1KB (should probably make it 2KB to be super safe).

Reply:
`maxChunkLen` here? instead of 1024?